### PR TITLE
Category and Relevance Classifier

### PR DESCRIPTION
**Pull request by georgerichardson**
- Downloads pretrained models from S3 bucket
- Contains all of the custom transformers and estimators needed
- If article is found to be relevant, calls `extract_reports` from `fact_extractor.py`

### CODE REVIEW

File: source/python/idetect/nlp_models/category/category.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                            f.write(chunk)
+                clf = joblib.load(default_model_path)
+        return clf
+
+    def predict(self, text):
+        try:
+            category = self.model.predict(pd.Series(text))[0]
+        except:
+            # if error occurs, classify as most likely category
+            category = 'disaster'
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: Returning a successful value in an error case strikes me as dangerous. What if there's a transient error that would be fixed by re-running? It seems like it would be better to let this exception bubble up and cause the Worker to mark this as failed, or to explicitly set a marker value that alerts a user to the fact that this failed.

File: source/python/idetect/nlp_models/category/category.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+class Tokenizer(TransformerMixin):
+    def __init__(self, stop_words=None):
+        self.stop_words = stop_words
+        self.tokenizer = WordPunctTokenizer()
+        
+    def prepare_tokens(self, text, stop_words):
+        tokens = self.tokenizer.tokenize(text)
+        tokens = [t for t in tokens if len(t) > 2]
+        tokens = [t for t in tokens if t not in stop_words]
+        tokens = [t.lower() for t in tokens]
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py:39]: Should this happen before the stop_words check? Are stop words intended to be case-sensitive?

File: source/python/idetect/nlp_models/category/category.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        self.model = self.load_model(model_path=model_path)
+
+    def load_model(self, model_path=None):
+        if model_path and os.path.isfile(model_path):
+            clf = joblib.load(model_path)
+        else:
+            default_model_path = 'category.pkl'
+            if os.path.isfile(default_model_path):
+                clf = joblib.load(default_model_path)
+            else:
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: What happens if several workers are started simultaneously, and they all check to see if the model is present, determine that it is not, and start downloading it? You're not locking the file that you're writing to below, so it's possible that the downloads will attempt to write at the same time and corrupt the file. Another possible failure is that one worker starts before another and starts downloading the file, and the other sees that the file exists and tries to read the only-partially-downloaded file.

Ideas:
1. Download to a tempfile (use the `tempfile` module, rather than a fixed file name for the same simultaneous write reason) and then move the file into place (`os.rename` is an atomic operation).
* Pros: easy to implement
* Cons: could be wasteful, as several downloads could take place
2. Lock the destination file on the filesystem. Note that you need to open the file and attempt to lock it this way before trying to read from it (the `load` case) as well as write it. It probably makes sense to do it first, and catch the file not found error in the case that it doesn't exist yet. 
```
import fcntl
... make sure the directory exists ...
with open(default_model_path, 'wb+') as f:
    try:
       fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB) # exclusive, nonblocking lock
       ... check if the file is empty, if not, someone else has already finished downloading it ...
       ... download and write the file ...
    catch BlockingIOError:
        # this happens if someone else is currently downloading it
        fcntl.flock(f, fcntl.LOCK_EX)  # exclusive, *blocking* lock. will wait until the other lock is released
    ...
    finally:
        fcntl.flock(x, fcntl.LOCK_UN) # probably not actually necessary because closing the filehandle will release the lock... I think.
```
* Pros: not wasteful
* Cons: trickier to implement correctly

File: source/python/idetect/nlp_models/category/category.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        self.no_above = no_above
+    
+    def make_dictionary(self, texts):
+        if not self.dictionary:
+            self.dictionary = gensim.corpora.Dictionary(texts)
+            if self.no_below or self.no_above:
+                self.dictionary.filter_extremes(no_below=self.no_below, no_above=self.no_above)
+        return self
+    
+    def tfidf_transform(texts, dictionary=None, tfidf_model=None):
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: Does this get called from anywhere? Why does it take dictionary as a parameter, rather than using `self.dictionary`?

File: source/python/idetect/nlp_models/category/category.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    
+    def lsi_to_vecs(self, corpus_lsi):
+        lsi_vecs = []
+        for c in corpus_lsi:
+            vec = [x[1] for x in c]
+            lsi_vecs.append(vec)
+        return np.array(lsi_vecs)
+    
+    def make_lsi_model(self, texts):    
+        self.corpus_tfidf, self.dictionary = self.build_tfidf(texts)
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: Will these ever be used again? I think they could just be local variables, rather than instance properties.

File: source/python/idetect/nlp_models/category/category.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        if not self.lsi_model:
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: Do you want to do this check _before_ calling `self.build_tfidf`? Is there value in calling that after there is already an LSI model?

File: source/python/idetect/nlp_models/category/category.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        return corpus_tfidf, dictionary, tfidf_model
+    
+    def make_corpus(self, texts):
+        corpus = [self.dictionary.doc2bow(text) for text in texts]
+        return corpus
+    
+    def make_tfidf_model(self, corpus):
+        if not self.tfidf_model:
+            self.tfidf_model = gensim.models.TfidfModel(corpus)
+        return self
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/category/category.py]: I don't think you need to return anything here. The caller must already have `self`, since it called a method on it.

Comment (WanderingStar) [reply source/python/idetect/nlp_models/category/category.py]: I think it's also a bit confusing that some of the `make_` methods return the thing they just made and others of them change the object. Should the mutators be called `set_tfidf_model`?

Also, it's surprising that a second call to `make_tfidf_model` with a different corpus silently does nothing. I would expect it to either replace the existing tfidf_model or to raise an exception to let the caller know that what they just tried didn't happen.

File: source/python/idetect/nlp_models/relevance/relevance.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+from sklearn.externals import joblib
+
+from idetect.model import Relevance
+
+class RelevanceModel(object):
+    def __init__(self, model_path=None):
+        self.model = self.load_model(model_path=model_path)
+
+    def load_model(self, model_path=None):
+        if model_path and os.path.isfile(model_path):
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/relevance/relevance.py]: This would make a great library function, so the code doesn't get repeated.

File: source/python/idetect/nlp_models/relevance/relevance.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        except:
+            # if error occurs, classify as most likely category
+            relevance = 'yes'
+
+        if relevance == 'yes':
+            return Relevance.DISPLACEMENT
+        else:
+            return Relevance.NOT_DISPLACEMENT
+
+class Tokenizer(TransformerMixin):
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/relevance/relevance.py]: Can this be declared in one place and imported here and in `category.py`?

File: source/python/idetect/nlp_models/relevance/relevance.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        self.tokenizer = WordPunctTokenizer()
+        self.stemmer = PorterStemmer()
+        
+    def prepare_stems(self, text, stop_words):
+        tokens = self.tokenizer.tokenize(text)
+        tokens = [t for t in tokens if len(t) > 2]
+        tokens = [t for t in tokens if t not in stop_words] 
+        stems = [self.stemmer.stem(t) for t in tokens]
+        stems = [s.lower() for s in stems]
+        stems = [s for s in stems if not s.isdigit()]
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/relevance/relevance.py]: Does `self.stemmer.stem(t)` return digits for tokens that weren't `isdigit()` to begin with? Maybe "1000s" => "1000"?

File: source/python/idetect/nlp_models/relevance/relevance.py (hunk 8 above)
Comment (WanderingStar) [reply source/python/idetect/nlp_models/relevance/relevance.py]: Or RelevanceModel and CategoryModel could both be instances of class DownloadableModel initialized with different `model_path` and `url` values.

File: source/python/idetect/nlp_models/relevance/relevance.py (hunk 9 above)
Comment (WanderingStar) [reply source/python/idetect/nlp_models/relevance/relevance.py]: Generally, I think we should repeat as little code as possible. If we find a bug in Tokenizer, we don't want to have to make sure to fix it in both places.

**Review by WanderingStar — CHANGES_REQUESTED** [review]

**Comment by WanderingStar** [comment]
Most of my comments are stylistic/DRY/optimizations, but I think the one about locking and the one about returning a non-error value in the case that an error occurs should be addressed.

File: source/python/idetect/nlp_models/category/category.py (hunk 2 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/category/category.py:39]: Yep, that's right. Swapped it.

File: source/python/idetect/nlp_models/category/category.py (hunk 4 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/category/category.py]: It was getting called in the LSI Transformer, but was unnecessary. Have removed.

File: source/python/idetect/nlp_models/category/category.py (hunk 7 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/category/category.py]: Think I've got most of these.

File: source/python/idetect/nlp_models/category/category.py (hunk 5 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/category/category.py]: These were a hangover from when I was testing the models, so I could inspect the attributes. Removed `self.corpus_tfidf`, but the dictionary does get used again.

File: source/python/idetect/nlp_models/category/category.py (hunk 6 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/category/category.py]: I actually removed this completely, it was just to accommodate a shortcut I was using when building models a while back.

File: source/python/idetect/nlp_models/relevance/relevance.py (hunk 8 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/relevance/relevance.py]: Made a base class `DownloadableModel`.

File: source/python/idetect/nlp_models/relevance/relevance.py (hunk 9 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/relevance/relevance.py]: Agree with this. Have a couple of thoughts to discuss on better ways to do it for the next PR.

File: source/python/idetect/nlp_models/relevance/relevance.py (hunk 10 above)
Comment (georgerichardson) [reply source/python/idetect/nlp_models/relevance/relevance.py]: Yeah, the stemmer does remove any characters appended to digits.

**Comment by georgerichardson** [comment]
Tried to address most of the comments above, but there's a few larger points I'll get a second opinion on in Slack before changing.

File: source/python/idetect/nlp_models/base_model.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                        raise
+                finally:
+                    fcntl.flock(f, fcntl.LOCK_UN)
+        else:  # download file
+            # if file directory doesn't exist create it
+            if os.path.dirname(model_path):
+                if not os.path.isdir(os.path.dirname(model_path)):
+                    try:
+                        os.makedirs(os.path.dirname(model_path))
+                    except OSError as exc:  # Guard against rare condition
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py]: "race" condition?

File: source/python/idetect/nlp_models/base_model.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+                        if exc.errno != errno.EEXIST:
+                            raise
+            with open(model_path, 'wb+') as f:
+                try:
+                    fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                    if os.path.getsize(model_path) == 0:
+                        r = requests.get(model_url, stream=True)
+                        for chunk in r.iter_content(chunk_size=1024):
+                            if chunk:  # filter out keep-alive new chunks
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py]: Out of curiosity, what happens if you call `f.write(chunk)` with an empty chunk?

File: source/python/idetect/nlp_models/base_model.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        # Load users model if specified and exists
+        if os.path.isfile(model_path):
+            with open(model_path, 'rb') as f:
+                # get exclusive lock in case currently being downloaded by
+                # another worker
+                try:
+                    fcntl.flock(f, fcntl.LOCK_EX)
+                    if os.path.getsize(model_path) > 0:
+                        model = joblib.load(model_path)
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py:44]: What happens in the else case here? What should happen? I assume we should download the model, right? Could you `return model` here and move the `else:` case of `if os.path.isfile(model_path):`, to the top level?

File: source/notebooks/file_download.ipynb [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [
+    {
```
Comment (WanderingStar) [inline source/notebooks/file_download.ipynb]: Would this be removed if we cleared the output in Jupyter?

**Comment by WanderingStar** [comment]
The locking logic looks pretty solid. It's the kind of thing I'd have to unit test to be sure of the edge cases, but I assume it generally works?

**Review by WanderingStar — APPROVED** [review]

**Review by simonb83 — COMMENTED** [review]
Just to understand how it will all fit together, which method will the Classification worker use?
